test_lbaas.py

```diff
@@ -0,0 +1,625 @@
+
+
+def test_algo_source_ip(prober, create_load_balancer_scenario, create_server,
+                        image):
+    """ The source_ip balancing algorithm always schedules connections from the
+    same source to the same pool member.
+
+    This test creates a load balancer with 4 backends and 4 clients with
+    different source IP addresses and verifies that connections from the same
+    client always go to the same backend.
+
+    """
+    # Create a load balancer setup with 4 backends on a private network
+    load_balancer, listener, pool, backends, private_network = \
+        create_load_balancer_scenario(
+            num_backends=4,
+            algorithm='source_ip',
+        )
+
+    # Create 4 client VMs
+    clients = in_parallel(
+        create_server,
+        instances=({
+            'name': f'client{i+1}',
+            'image': image,
+        } for i in range(4))
+    )
+
+    # Build map of which client is directed to which backend server
+    backend_per_client = {}
+    for client in clients:
+        # Issue the first request from this client to the LB
+        backend_per_client[client.name] = load_balancer.get_url(
+            client,
+            url='/hostname',
+        )
+
+    # Issue 10 requests and verify the correct backend receives the request
+    for n in range(10):
+        for client in clients:
+            backend_hit = load_balancer.get_url(client, url='/hostname')
+            assert backend_per_client[client.name] == backend_hit
+
+
+def test_algo_least_connections(server, create_load_balancer_scenario):
+    """ The least_connections balancing algorithm schedules connections to the
+    backend with the least amount of active connections.
+
+    Note: The current connection count is reset on any configuration changes to
+          the load balancer. Still running connections initiated before the
+          configuration change are not considered for selecting the backend.
+    """
+
+    # This test uses a function scoped prober because we are going to start
+    # long running downloads which should not impact other tests.
+    prober = server
+
+    # Create a load balancer setup with 2 backends on a private network
+    load_balancer, listener, pool, backends, private_network = \
+        create_load_balancer_scenario(
+            num_backends=2,
+            algorithm='least_connections',
+        )
+
+    # start a persistent endless download of random data to "block" one backend
+    backend_endless = start_persistent_download(prober, load_balancer,
+                                                backends)
+
+    # Verify requests go to the other backend as it has less active
+    # connections
+    for i in range(10):
+        backend_hit = load_balancer.get_url(prober, url='/hostname')
+
+        assert backend_hit != backend_endless.name
+
+
+@pytest.mark.parametrize('health_monitor',
+                         [{'type': 'ping',
```

Comment:
I find this somewhat ugly, can we move this into a global function at the top of the file? e.g. `HEALTH_MONITOR_TYPES`. Also, it would be nice if we could just generate the ids, instead of having to basically copy the `type`.

Reply:
I moved the http configuration options into a dictionary inside the function. I would prefer not to move this to the top of the file to keep everything at one place. I think this piece is relevant to understand the test.